### PR TITLE
The Wettest Update (1.13)

### PR DESCRIPTION
**Pull request by mastercoms**
Glowstone is proud to announce the Update Aquatic (well, Mojang is, we're pretty panicked right now)

#### The big one
- [ ] [1.13 API](https://www.spigotmc.org/threads/bukkit-craftbukkit-spigot-bungeecord-1-13-pre7-development-builds.328883/) compatibility, [Summary](https://www.spigotmc.org/threads/minecraft-1-13-what-to-expect.288714/)
  - [ ] Bukkit [1.13 update](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/f8b2086d60942eb2cd7ac25a2a1408cb790c222c)

#### The other big one
- [ ] [1.13 Protocol compatibility](https://wiki.vg/Protocol) (#958)

#### Feature completeness
So. Many. Features. This list is WIP and will only cover major feature additions/changes. If you are a developer, feel free to edit and add more items and detail to this list. For a comprehensive list of changes, refer to [the wiki](https://minecraft.gamepedia.com/1.13).

Most features in this list are non-blocking to this PR getting merged, but they'll be nice to implement. However, there are a few things that have to be implemented or else we wouldn't really be 1.13 (mostly just the stuff in the `General` section below).

**General**
- [ ] Data packs support

Can I interrupt this list for a second to just say how _great_ data packs are going to be for Glowstone? They solve a ton of issues with us not being able to replicate game data from vanilla. Ok, let's carry on.

- [ ] Tags
- [ ] Where did our IDs go? (also known as "the flattening") (#960)

**Blocks**
- [ ] Different air types
- [ ] Stripped logs
- [ ] Stripped wood
- [ ] Turtle Egg
- [ ] New water physics behavior

**Items**
- [ ] Debug stick
- [ ] New recipes _TODO_
  - [ ] Wood block
- [ ] Trident weapon
- [ ] New spawn eggs _TODO_

**World**
- [ ] Coral reefs
- [ ] Icebergs
- [ ] Shipwrecks
- [ ] Underwater caves
- [ ] Underwater ruins
- [ ] Kelp
- [ ] Bubble columns
- [ ] Seagrass
- [ ] Sea Pickle
- [ ] Biomes (https://minecraft.gamepedia.com/Java_Edition_1.13-pre5#Changes)
  - [ ] The End
    - [ ] Small End Islands
    - [ ] End Midlands
    - [ ] End Highlands
    - [ ] End Barrens
  - [ ] Ocean
    - [ ] Warm Ocean
    - [ ] Lukewarm Ocean
    - [ ] Cold Ocean
    - [ ] Deep Warm Ocean
    - [ ] Deep Lukewarm Ocean
    - [ ] Deep Warm Ocean
    - [ ] Deep Cold Ocean
    - [ ] Deep Frozen Ocean
    - [ ] Frozen Ocean
- [ ] Buffet world type

**Status Effects/Potions**
- [ ] Slow falling
- [ ] Turtle master
- [ ] Status: Conduit Power
- [ ] Status: Dolphin's Grace

**Enchantments**
- [ ] Channelling
- [ ] Impaling
- [ ] Loyalty
- [ ] Riptide

**Commands**
- [ ] /bossbar
- [ ] /data
- [ ] /datapack
- [ ] /locate
- [ ] /scoreboard render types hearts and integer added
- [ ] /tp an alias of the new /teleport
- [ ] /time supports setting to noon and midnight

More changes listed here: https://minecraft.gamepedia.com/1.13#Command_format_2

**Advancements**

**Entities**
- [ ] Dolphins
- [ ] Drowned
- [ ] Fish _TODO_
  - [ ] Cod
  - [ ] Salmon
  - [ ] Puffer Fish
  - [ ] Tropical Fish
- [ ] Phantoms
- [ ] Turtles
- [ ] Trident (projectile)

### CODE REVIEW

File: src/main/java/net/glowstone/entity/GlowEntity.java [hunk 1]
```diff
@@ -697,9 +697,16 @@ public void setRawLocation(Location location, boolean fall) {
         if (hasMoved()) {
             if (!fall || type == Material.LADDER // todo: horses are not affected
                     || type == Material.VINE // todo: horses are not affected
```
Comment (Pr0methean) [inline src/main/java/net/glowstone/entity/GlowEntity.java:726]: `this instanceof Horse` (or some subtype) `&&`

File: src/main/java/net/glowstone/entity/GlowEntity.java [hunk 2]
```diff
@@ -1487,6 +1494,18 @@ public boolean entityInteract(GlowPlayer player, InteractEntityMessage message)
         return false;
     }
 
+    @Override
+    public boolean isPersistent() {
+        // TODO: 1.13
+        throw new UnsupportedOperationException("Not supported yet.");
```
Comment (Pr0methean) [inline src/main/java/net/glowstone/entity/GlowEntity.java:1562]: Just add `@Getter @Setter private boolean persistent` as a starting point, and move the TODO there.

File: src/main/java/net/glowstone/entity/GlowLivingEntity.java [hunk 3]
```diff
@@ -1243,6 +1224,24 @@ public boolean isHandRaised() {
         return false;
     }
 
+    @Override
+    public boolean isSwimming() {
+        // TODO: 1.13
+        throw new UnsupportedOperationException("Not implemented yet.");
```
Comment (Pr0methean) [inline src/main/java/net/glowstone/entity/GlowLivingEntity.java:1284]: Add `@Getter @Setter private boolean swimming` as a starting point, and move the TODO there.

File: src/main/java/net/glowstone/entity/objects/GlowArmorStand.java [hunk 4]
```diff
@@ -277,26 +280,31 @@ private EquipmentSlot getEquipType(Material mat) {
             case IRON_HELMET:
```
Comment (Pr0methean) [inline src/main/java/net/glowstone/entity/objects/GlowArmorStand.java]: What about the turtle-shell helmet?

File: src/main/java/net/glowstone/entity/GlowLivingEntity.java (hunk 3 above)
Comment (aramperes) [reply src/main/java/net/glowstone/entity/GlowLivingEntity.java:1284]: I'd rather have this error thrown if a plugin uses unimplemented APIs than failing silently. At least then, incompatibility is made clear to the end user.

File: src/main/java/net/glowstone/entity/objects/GlowArmorStand.java (hunk 4 above)
Comment (aramperes) [reply src/main/java/net/glowstone/entity/objects/GlowArmorStand.java]: We should probably group equipment types anyway (or is that already done?). The turtle-shell can be added to such a group.

File: src/main/java/net/glowstone/entity/GlowEntity.java (hunk 1 above)
Comment (aramperes) [reply src/main/java/net/glowstone/entity/GlowEntity.java:726]: I'm confused about this todo, does it mean horses should take fall damage even on these blocks, or should be nullified entirely for all horses?

File: src/main/java/net/glowstone/entity/GlowEntity.java (hunk 2 above)
Comment (aramperes) [reply src/main/java/net/glowstone/entity/GlowEntity.java:1562]: See the other comment on this

File: src/main/java/net/glowstone/entity/objects/GlowArmorStand.java (hunk 4 above)
Comment (Pr0methean) [reply src/main/java/net/glowstone/entity/objects/GlowArmorStand.java]: I'd suggest, during the next change to Glowkit, adding the equipment slot as a field of Material with a getter.

**Comment by WesleyVanNeck** [comment]
Rip 1.14.1 is out

**Comment by mastercoms** [comment]
Superseded by #1091

**Pull request closed without merging**